### PR TITLE
Feature/error management

### PR DESCRIPTION
**Pull request by thibautbenjamin**
Improve error reporting so that catt does not crash when encountering an error

### CODE REVIEW

File: bin/catt.ml [hunk 1]
```diff
@@ -13,20 +13,23 @@ let parse_file f =
...
+      "--no-builtins", Stdlib.Arg.Set no_builtins, "Prevent using built-in compositions and identities";
+      "--debug", Stdlib.Arg.Set debug, "Debug mode: stop on error and drops a menu"
     ]
     (fun s -> file_in := s::!file_in)
     usage;
   let _ = match !file_in with
     | [f] ->
       if !no_builtins then Catt.Settings.use_builtins := false;
-      Catt.Command.exec (parse_file f)
+      if !debug then Catt.Settings.debug := true;
```
Comment (regular-citizen) [inline bin/catt.ml:32]: Can this be replaced with '''Catt.Settings.debug = !debug'''? Also why are these flags negated?

File: lib/io.ml [hunk 2]
```diff
@@ -5,14 +5,11 @@ let print_endline s = print_string s; print_newline ()
...
 
-let printf e = Printf.ksprintf print_string e
-let debug e =
-  Printf.ksprintf
-    (fun s -> printf "=D.D= %s.\n\n%!" s) e
+let printf e = Printf.printf e
+let eprintf e = Printf.eprintf e
+let debug e = Printf.ksprintf (fun s -> printf "[=D.D=] %s\n\n%!" s) e
 let info ?(v = 0) e =
   if !Settings.verbosity >= v then
```
Comment (regular-citizen) [inline lib/io.ml:12]: Replace with '''Settings.verbosity <= v''

File: lib/kernel.ml [hunk 3]
```diff
@@ -854,3 +876,27 @@ end = struct
...
+  | NotEqual(s1,s2) ->
+    Error.untypable name (Printf.sprintf "%s and %s are not equal" s1 s2)
+  | InvalidSubTarget (s,tgt) ->
+    Error.untypable name (Printf.sprintf "substitution %s does not apply from context %s" s tgt)
+  | Error.UnknownId (s) ->
+    Error.untypable name (Printf.sprintf "unknown identifier :%s" s)
+  | MetaVariable ->
+    Error.incomplete_constraints name
+
+let check_type ctx a =
```
Comment (regular-citizen) [inline lib/kernel.ml:891]: Potential performance issue here: these error strings are always built when checking even though check_fn need not fail

File: lib/prover.ml [hunk 4]
```diff
@@ -6,10 +6,12 @@ let parse s =
   with
   | Failure s when s = "lexing: empty token" ->
      let pos = Lexing.lexeme_end_p lexbuf in
-     Io.error "lexing error in file %s at line %d, character %d"
-     pos.Lexing.pos_fname
-     pos.Lexing.pos_lnum
-     (pos.Lexing.pos_cnum - pos.Lexing.pos_bol)
+     failwith
+       (Printf.sprintf
+          "lexing error in file %s at line %d, character %d"
```
Comment (regular-citizen) [inline lib/prover.ml:11]: We could print the relevant fragment of the file at this point to give the user a better idea of where lexing failed

File: bin/catt.ml (hunk 1 above)
Comment (thibautbenjamin) [reply bin/catt.ml:32]: Oh no, I did what I keep telling the students not to do...

Also: weird OCaml quirk: `!` is not the negation, it is the operation of accessing the value stored in a reference. Negation is written as `not` :)

File: lib/io.ml (hunk 2 above)
Comment (thibautbenjamin) [reply lib/io.ml:12]: Again, `!` is not the negation

File: lib/kernel.ml (hunk 3 above)
Comment (thibautbenjamin) [reply lib/kernel.ml:891]: Interesting point, and I definitely had performances issues in the past due to the computing strings that were never printed. But in this case, I doubt it happens: For instance how would you print the string corresponding to the `untypable` error without actually catching the exception, you don't know the values of `s1` and `s2` beforehand

File: lib/prover.ml (hunk 4 above)
Comment (thibautbenjamin) [reply lib/prover.ml:11]: The management of lexing and parsing is still terrible, and I agree that would be a nice thing to have in general. However this information is not accessible directly here and would require a complete refactoring. I propose that we defer refactoring the handling of lexing/parsing errors for another MR, as I believe those are less critical

**Comment by thibautbenjamin** [comment]
Replied to all your comments and fixed a few things, should be ready to merge now :)

File: lib/prover.ml (hunk 4 above)
Comment (thibautbenjamin) [reply lib/prover.ml:11]: We should open an issue about that upon merging so that we don't forget

File: lib/kernel.ml (hunk 3 above)
Comment (regular-citizen) [reply lib/kernel.ml:891]: My main concern are not the sprintf in check, it's the computation of '''name'''. That's not lazily evaluated, that's computed every time a check function is called and silently prints a whole type/term to string.

File: lib/prover.ml (hunk 4 above)
Comment (regular-citizen) [reply lib/prover.ml:11]: Do we not have access to lexbuf here? I agree that it is not immediate by the way.

**Comment by regular-citizen** [comment]
Added a couple replies, we're very close to merging tho

File: lib/kernel.ml (hunk 3 above)
Comment (thibautbenjamin) [reply lib/kernel.ml:900]: Oh yes, good catch. Sorry I had misunderstood, I made the evaluation lazy. I had to do something a bit  hackish to be able to lazily evaluate, but still reuse the result of the evaluation when it has already been computed

File: lib/prover.ml (hunk 4 above)
Comment (thibautbenjamin) [reply lib/prover.ml:11]: We do have access to lexbuf at this point. To be honest, I haven't touched or looked at this part of the code in years, and there is definitely a significant amount of work to handle lexing and parsing errors properly. As of right now, I believe it is really broken. Adding the info you suggested would be a good idea if the code was doing what it is supposed to... But given it's all broken, we should fix it first and then implement these kinds of nice quality of life features :)

**Comment by thibautbenjamin** [comment]
pushed a fix

**Pull request merged**